harness/tests/experiment/pytorch/test_deepspeed_autotuning.py

```diff
@@ -134,75 +170,67 @@ def _process_operations(self, operations: List[searcher.Operation]) -> None:
             self._append_events_for_op(op)  # validate_after returns two events.
 
     def handle_get_events(self) -> Optional[Sequence[bindings.v1SearcherEvent]]:
-        return self.events_queue
+        return list(self.events_queue)
 
     def _append_events_for_op(self, op: searcher.Operation) -> None:
         if isinstance(op, searcher.ValidateAfter):
             index = min(self.metric_index, len(self.all_metrics) - 1)
             metric = self.all_metrics[index]
-            validation_completed = bindings.v1ValidationCompleted(
-                requestId=str(op.request_id),
-                metric=metric,
-                validateAfterLength=str(op.length),
-            )
             self.metric_index += 1
-            self.events_count += 1
-            event = bindings.v1SearcherEvent(
-                id=self.events_count, validationCompleted=validation_completed
-            )
-            self.events_queue.append(event)
-
-            trial_progress = bindings.v1TrialProgress(
-                requestId=str(op.request_id), partialUnits=float(op.length)
-            )
-            self.events_count += 1
-            event = bindings.v1SearcherEvent(id=self.events_count, trialProgress=trial_progress)
-            self.events_queue.append(event)
-
-        if isinstance(op, searcher.Create):
+            if ERROR_METRIC_NAME in metric:
+                trial_exited_early = bindings.v1TrialExitedEarly(
+                    requestId=str(op.request_id),
+                    exitedReason=bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_UNSPECIFIED,
+                )
+                self.events_count += 1
+                event = bindings.v1SearcherEvent(
+                    id=self.events_count, trialExitedEarly=trial_exited_early
+                )
+                self.events_queue.append(event)
+
+                trial_closed = bindings.v1TrialClosed(requestId=str(op.request_id))
+                self.events_count += 1
+                event = bindings.v1SearcherEvent(id=self.events_count, trialClosed=trial_closed)
+                self.events_queue.append(event)
+            else:
+                validation_completed = bindings.v1ValidationCompleted(
+                    requestId=str(op.request_id),
+                    metric=metric,
+                    validateAfterLength=str(op.length),
+                )
+
+                self.events_count += 1
+                event = bindings.v1SearcherEvent(
+                    id=self.events_count, validationCompleted=validation_completed
+                )
+                self.events_queue.append(event)
+
+                # Send 1.0 to signal it was completed
+                trial_progress = bindings.v1TrialProgress(
+                    requestId=str(op.request_id), partialUnits=1.0
```

Comment:
This `partialUnits` bit fixes where we previously saw `6.0`?

Reply:
yeah, I wasn't really sure how we wanted to handle this or whether the partialUnits was even relevant in our use case

Reply:
I'm not actually sure how `v1TrialProgress` is intended to be used, particularly the `partialUnits` arg. Hardly gets used at all in the codebase, actually.